### PR TITLE
feat: mcp resources

### LINKED ISSUE #316: Add MCP resource tools to Template Agent
### Problem

Template Agent can call MCP server **tools** but has no way to read MCP **resources** (`resources/list`, `resources/templates/list`, `resources/read`). The LLM cannot speak JSON-RPC, so resources on any connected MCP server are invisible to the agent.

### Proposed solution

Three host-side LangChain tools that proxy MCP `resources/list`, `resources/templates/list`, and `resources/read`.
`mcp_read_resource` extracts `contents[].text`, paginates by line (`offset`/`limit`, default 100), char-caps at 400k, cuts at line boundaries, and pre-splits long single lines. Blobs are stubbed. Added to the eviction skip list so results stay inline — no `/large_tool_results/` detour.
Scoped by `resources:` frontmatter allowlist. Subagents inherit the orchestrator's list. Graph cache busts on allowlist change. Fully stateless — no process-local state, multi-pod safe.

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AtrikGhosh**
## What

Add spec-compliant MCP resource tools (`mcp_list_resources`, `mcp_list_resource_templates`, `mcp_read_resource`) so agents can read resources from any enabled MCP server, with `read_file`-style pagination and context-safe defaults.

Fixes #316

## How

**Three new LangChain tools** in `mcp_resource_tools.py` that proxy MCP `resources/list`, `resources/templates/list`, and `resources/read` at turn time. No catalog cache — stateless and multi-pod safe.

**Pagination** on `mcp_read_resource`: `offset`/`limit` (defaults 0/100), 400k char budget (sized for full GUIDANCE.md), line-boundary truncation, long-line pre-splitting. Blobs replaced with `[Binary content omitted (mime)]` stubs.

**Eviction skip list**: `mcp_read_resource` added to `TOOLS_EXCLUDED_FROM_EVICTION` so FilesystemMiddleware does not offload results to `/large_tool_results/`.

**Scoping**: `resources:` frontmatter allowlist (URI / `uriTemplate` strings). Omit or `[]` = all URIs. Subagents without `resources:` inherit the orchestrator's list.

**Graph cache**: fingerprint now includes `resource_uris` and `mcp_server_names` so allowlist changes bust the cache.

**Config validation**: `loader.py` validates `resources:` as a list of strings at load time.

## Testing

- [x] Unit tests added/updated (35 resource tool tests, 5 graph tests, 6 subagent tests, config validation tests)
- [x] Ran locally (`uv run pytest tests/unit -x`)
- [x] Manual verification against mock MCP server: pagination (short/long/wide/huge-line), blob handling, template URI matching, server/URI allowlist rejection, edge cases (negative limit, offset past end, blank-line + long-line), resource inheritance (orchestrator restricts → subagent inherits → subagent overrides)

## Rollback

Revert the three commits. Resource tools are additive — removing them leaves agents with only MCP server tools, which is the pre-existing behavior.

## Checklist

- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) (`feat:`, `fix:`, `ci:`, etc.)
- [x] No secrets, credentials, or PII in the diff
- [x] No breaking changes (or documented above with a migration path)
- [x] Pre-commit hooks pass (`uv run pre-commit run --all-files`)

### CODE REVIEW

File: deep_agent/aegra/mcp_resource_tools.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+        and isinstance(detail.get("mcp_name"), str)
+        and isinstance(detail.get("connect_url"), str)
+    )
+
+
+def _raise_or_format_http(exc: HTTPException) -> str:
+    if _is_authorization_required(exc):
+        detail = exc.detail
+        raise NeedsAuthorization(detail["mcp_name"], detail["connect_url"]) from None
+    return f"MCP resource request failed ({exc.status_code}): {exc.detail}"
```
[the remark this answers is not in the document]
Comment (AtrikGhosh) [reply deep_agent/aegra/mcp_resource_tools.py:221]: Skipped. authorization_required already becomes NeedsAuthorization and the connect interrupt. Other HTTP errors here are not all “api_key missing” — they include 404 (unknown/disabled server), 401 (no SSO bearer), and 400 (bad URI). Mapping every one of those to an api_key config message would be wrong and would hide a useful status from the model. Raw TLS/host dumps are already stripped on the generic Exception path via _format_generic_failure. These HTTP details are short, controlled strings we raise on purpose.

File: tests/unit/aegra/test_graph.py [hunk 2]
```diff
@@ -606,6 +632,174 @@ async def test_returns_cached_graph_on_hit(self):
...
+        mock_config.get_mcp_servers.return_value = {
+            "template-mcp-server": {"enabled": True},
+        }
+
+        _result, _compiled, mock_create, mock_mw, _subs = await self._build_agent(
+            mock_config
+        )
+
+        names = [t.name for t in mock_create.call_args.kwargs["tools"]]
+        assert LIST_TOOL in names
```
[the remark this answers is not in the document]
Comment (AtrikGhosh) [reply tests/unit/aegra/test_graph.py]: Fixed

File: deep_agent/aegra/mcp_resource_tools.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            args_schema=_ListInput,
+        ),
+        StructuredTool(
+            name=READ_TOOL,
+            description=read_desc,
+            func=lambda **_: "",
+            coroutine=_read,
+            args_schema=_ReadInput,
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (AtrikGhosh) [reply deep_agent/aegra/mcp_resource_tools.py:415]: Fixed

File: tests/unit/aegra/test_mcp_resource_tools.py [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+            result = await _tool(tools, READ_TOOL).ainvoke(
+                {"mcp_name": "s", "uri": "template://mixed"}
+            )
+        content_before_notice = result.split("\n\n[Output truncated")[0]
+        assert content_before_notice.endswith("\n")
+        assert "offset=" in result
+        offset_val = int(result.split("offset=")[1].split(" ")[0].rstrip(","))
+        assert offset_val < 100
+        assert "short-0" in result
+        assert f"short-{offset_val}" not in content_before_notice
```
[the remark this answers is not in the document]
Comment (AtrikGhosh) [reply tests/unit/aegra/test_mcp_resource_tools.py]: fixed

**Comment by NP-compete** [comment]
@coderabbitai review

**Review by NP-compete — APPROVED** [review]
